pkg/placement/membership.go

```diff
@@ -131,86 +133,112 @@ func (p *Service) membershipChangeWorker(stopCh chan struct{}) {
 			disseminateTimer.Stop()
 			return
 
-		case op := <-p.membershipCh:
-			p.processRaftStateCommand(op)
-
-		case <-disseminateTimer.C:
+		case t := <-disseminateTimer.C:
 			// check if there is actor runtime member change.
-			if p.memberUpdateCount > 0 {
-				log.Debugf("request dessemination. memberUpdateCount count: %d", p.memberUpdateCount)
-				p.membershipCh <- hostMemberChange{cmdType: raft.TableDisseminate}
+			if p.disseminateNextTime <= t.UnixNano() && len(p.membershipCh) == 0 {
+				cnt := p.memberUpdateCount.Load()
+				if cnt > 0 {
+					log.Debugf("Add raft.TableDisseminate to membershipCh. memberUpdateCount count: %d", cnt)
+					p.membershipCh <- hostMemberChange{cmdType: raft.TableDisseminate}
+				}
 			}
 
 		case t := <-faultyHostDetectTimer.C:
 			// Each dapr runtime sends the heartbeat every one second and placement will update UpdatedAt timestamp.
 			// If UpdatedAt is outdated, we can mark the host as faulty node.
 			// This faulty host will be removed from membership in the next dissemination period.
-			m := p.raftNode.FSM().State().Members
-			for _, v := range m {
-				if t.Sub(v.UpdatedAt) <= p.faultyHostDetectDuration {
-					continue
-				}
-				log.Debugf("try to remove outdated hosts: %s, elapsed: %d ms", v.Name, t.Sub(v.UpdatedAt).Milliseconds())
-
-				p.membershipCh <- hostMemberChange{
-					cmdType: raft.MemberRemove,
-					host:    raft.DaprHostMember{Name: v.Name},
+			if len(p.membershipCh) == 0 {
+				m := p.raftNode.FSM().State().Members
+				for _, v := range m {
+					heartbeat, ok := p.lastHeartBeat.Load(v.Name)
+					if !ok {
+						p.lastHeartBeat.Store(v.Name, time.Now().UnixNano())
```

Comment:
a new host with no heartbeat record will not be removed till 2*faultyHostDetectTimer?

Reply:
yes. but I added more comment. This will happen only when leadership is changed. please see the code comment.